### PR TITLE
Handle implementation organization

### PR DESCRIPTION
**Pull request by jblebrun**
* Move single,collection implementation into separate files
* Simplify implementation strategy: always instantiate full handle, wrap
  in object that only exposes proper interface.
* Remove dependencies on storage.handle (including moving
  RawEntityDereferencer to different package, since we still need it

### CODE REVIEW

**Comment by jblebrun** [comment]
These changes made #5068 a fair bit easier

File: java/arcs/core/host/EntityHandleManager.kt [hunk 1]
```diff
@@ -79,53 +79,22 @@ class EntityHandleManager(
...
+                name = idGenerator.handleName(baseName),
                 entitySpec = entitySpec,
-                storageProxy = storageProxy,
-                entityPreparer = entityPrep,
+                storageProxy = singletonStoreProxy(storageKey, entitySpec),
+                entityPreparer = EntityPreparer(name, idGenerator, entitySpec.SCHEMA, ttl, time),
                 dereferencerFactory = dereferencerFactory
             )
+        }.let { handle ->
+            return when (mode) {
```
Comment (cromwellian) [inline java/arcs/core/host/EntityHandleManager.kt]: Nice. Although we've come full circle back to facades again, this trick is a lot similar the original one I was doing in BaseParticle and my original PR.

File: java/arcs/core/entity/CollectionHandle.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    // endregion
+
+    // region implement WriteCollectionHandle<T>
+    override suspend fun store(entity: T) {
+        storageProxy.applyOp(
+            CrdtSet.Operation.Add(
+            name,
+            storageProxy.getVersionMap().increment(name),
+            entityPreparer.prepareEntity(entity)
+        ))
```
Comment (jasonwyatt) [inline java/arcs/core/entity/CollectionHandle.kt]:
```suggestion
        storageProxy.applyOp(
            CrdtSet.Operation.Add(
                name,
                storageProxy.getVersionMap().increment(name),
                entityPreparer.prepareEntity(entity)
            )
        )
```

File: java/arcs/core/entity/CollectionHandle.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    override suspend fun clear() {
+        storageProxy.getParticleView().forEach {
+            storageProxy.applyOp(
+                CrdtSet.Operation.Remove(
+                name,
+                storageProxy.getVersionMap(),
+                it
+            ))
```
Comment (jasonwyatt) [inline java/arcs/core/entity/CollectionHandle.kt]:
```suggestion
            storageProxy.applyOp(
                CrdtSet.Operation.Remove(
                    name,
                    storageProxy.getVersionMap(),
                    it
                )
            )
```

File: java/arcs/core/entity/CollectionHandle.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+
+    override suspend fun remove(entity: T) {
+        storageProxy.applyOp(
+            CrdtSet.Operation.Remove(
+            name,
+            storageProxy.getVersionMap(),
+            entity.serialize()
+        ))
```
Comment (jasonwyatt) [inline java/arcs/core/entity/CollectionHandle.kt]:
```suggestion
        storageProxy.applyOp(
            CrdtSet.Operation.Remove(
                name,
                storageProxy.getVersionMap(),
                entity.serialize()
            )
        )
```

File: java/arcs/core/entity/CollectionHandle.kt [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    override suspend fun createReference(entity: T): Reference<T> {
+        val entityId = requireNotNull(entity.entityId) {
+            "Entity must have an ID before it can be referenced."
+        }
+        val storageKey = requireNotNull(storageProxy.storageKey as? ReferenceModeStorageKey) {
+            "ReferenceModeStorageKey required in order to create references."
+        }
+        if (!fetchAll().any { it.entityId == entityId }) {
+            throw IllegalArgumentException("Entity is not stored in the Collection.")
+        }
```
Comment (jasonwyatt) [inline java/arcs/core/entity/CollectionHandle.kt]: This can be done with a `require()` call.

File: java/arcs/core/entity/CollectionHandle.kt [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        return Reference(
+            entitySpec,
+            arcs.core.storage.Reference(entity.serialize().id, storageKey.backingKey, null).also {
+                it.dereferencer = dereferencerFactory.create(entitySpec.SCHEMA)
+            }
+        )
+    }
+    // endregion
+
+    private fun adaptValues(values: Set<RawEntity>) = values.mapTo(mutableSetOf()) {
```
Comment (jasonwyatt) [inline java/arcs/core/entity/CollectionHandle.kt]: Does this mean it returns a MutableSet?

File: java/arcs/core/entity/EntityPreparer.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ */
+package arcs.core.entity
+
+import arcs.core.common.Id
+import arcs.core.data.RawEntity
+import arcs.core.data.Schema
+import arcs.core.data.Ttl
+import arcs.core.util.Time
+
+/** Prepares the provided entity and serializes it for storage */
```
Comment (jasonwyatt) [inline java/arcs/core/entity/EntityPreparer.kt]:
```suggestion
/** Prepares the provided entity and serializes it for storage. */
```

File: java/arcs/core/entity/SingletonHandle.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import arcs.core.storage.StorageProxy
+import arcs.core.storage.StoreOptions
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+
+typealias SingletonProxy<T> = StorageProxy<SingletonData<T>, SingletonOp<T>, T?>
+typealias SingletonData<T> = CrdtSingleton.Data<T>
+typealias SingletonOp<T> = CrdtSingleton.IOperation<T>
+typealias SingletonStoreOptions<T> = StoreOptions<SingletonData<T>, SingletonOp<T>, T?>
+
+class SingletonHandle<T : Entity>(
```
Comment (jasonwyatt) [inline java/arcs/core/entity/SingletonHandle.kt]: KDoc would be helpful here too.

File: java/arcs/core/entity/SingletonHandle.kt [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            entityPreparer.prepareEntity(entity)
+        ))
+    }
+
+    override suspend fun clear() {
+        storageProxy.applyOp(
+            CrdtSingleton.Operation.Clear(
+            name,
+            storageProxy.getVersionMap()
+        ))
```
Comment (jasonwyatt) [inline java/arcs/core/entity/SingletonHandle.kt]:
```suggestion
        storageProxy.applyOp(
            CrdtSingleton.Operation.Clear(
                name,
                storageProxy.getVersionMap()
            )
        )
```

File: java/arcs/core/entity/SingletonHandle.kt [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    // endregion
+
+    // region implement WriteSingletonHandle<T>
+    override suspend fun store(entity: T) {
+        storageProxy.applyOp(
+            CrdtSingleton.Operation.Update(
+            name,
+            storageProxy.getVersionMap().increment(name),
+            entityPreparer.prepareEntity(entity)
+        ))
```
Comment (jasonwyatt) [inline java/arcs/core/entity/SingletonHandle.kt]:
```suggestion
        storageProxy.applyOp(
            CrdtSingleton.Operation.Update(
                name,
                storageProxy.getVersionMap().increment(name),
                entityPreparer.prepareEntity(entity)
            )
        )
```

File: java/arcs/core/entity/SingletonHandle.kt [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+    override suspend fun createReference(entity: T): Reference<T> {
+        val entityId = requireNotNull(entity.entityId) {
+            "Entity must have an ID before it can be referenced."
+        }
+        val storageKey = requireNotNull(storageProxy.storageKey as? ReferenceModeStorageKey) {
+            "ReferenceModeStorageKey required in order to create references."
+        }
+        if (fetch()?.entityId != entityId) {
```
Comment (jasonwyatt) [inline java/arcs/core/entity/SingletonHandle.kt]: Also could be done with `require`

File: java/arcs/core/entity/CollectionHandle.kt [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+import arcs.core.storage.StorageProxy
+import arcs.core.storage.StoreOptions
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+
+typealias CollectionData<T> = CrdtSet.Data<T>
+typealias CollectionOp<T> = CrdtSet.IOperation<T>
+typealias CollectionStoreOptions<T> = StoreOptions<CollectionData<T>, CollectionOp<T>, Set<T>>
+typealias CollectionProxy<T> = StorageProxy<CrdtSet.Data<T>, CrdtSet.IOperation<T>, Set<T>>
+
+class CollectionHandle<T : Entity>(
```
[the remark this answers is not in the document]
Comment (jblebrun) [reply java/arcs/core/entity/CollectionHandle.kt]: good call

File: java/arcs/core/entity/CollectionHandle.kt (hunk 6 above)
Comment (jblebrun) [reply java/arcs/core/entity/CollectionHandle.kt]: Yes, will fix.

File: java/arcs/core/entity/CollectionHandle.kt [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+import arcs.core.storage.StoreOptions
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+
+typealias CollectionData<T> = CrdtSet.Data<T>
+typealias CollectionOp<T> = CrdtSet.IOperation<T>
+typealias CollectionStoreOptions<T> = StoreOptions<CollectionData<T>, CollectionOp<T>, Set<T>>
+typealias CollectionProxy<T> = StorageProxy<CrdtSet.Data<T>, CrdtSet.IOperation<T>, Set<T>>
+
+/**
+ * This class implements all of the methods that are needed by the various singleton [Handle]
```
Comment (csilvestrini) [inline java/arcs/core/entity/CollectionHandle.kt]: singleton -> collection

**Pull request merged**